### ISSUE TITLE
HTML color codes surrounded by spaces are not considered valid

### ISSUE THREAD

**Issue by guemax** [godotengine/godot#79338]
### Godot version

v4.1.stable.official [970459615]

### System information

Godot v4.1.stable - Pop!_OS 22.04 LTS - Vulkan (Compatibility) - NVIDIA GeForce MX350 (nvidia; 525.116.04) - 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz (8 Threads)

### Issue description

When I use the color picker to change a color, I can paste an HTML color code or write it myself. I can also prefix my text with a `# &mdash; the color picker will accept it as a valid color and update the corresponding element with that color. This is especially useful when I copy an HTML color code from a color palette like the one on lospec.com, since they always have a hash in front of them.

However, I cannot paste HTML color codes with surrounding spaces into this text box, even though the color code (without the spaces) is valid.

I would definitely call this behaviour a bug, since other programs like Affinity Designer, Blender or Gimp are very permissive about the input of HTML color codes. Or according to Jon Postel, "Be conservative in what you send, be liberal in what you accept.".

### Steps to reproduce

Preparation: 

1. create a new Godot project
2. create a new 2d scene
3. add a ColorRect node.
4. open the color picker for the color attribute in the inspector tab.

Explore the bug:

1. enter a valid HTML color code like `000` (black) => color is updated.
2. enter a valid HTML color code like `#fff` (white) => color is updated.
3. enter a valid HTML color code with leading space, for example `   #000` => color is not updated.
4. enter a valid HTML color code like `#000`(black) => color is updated.
5. enter a valid HTML color code with surrounding spaces, e.g. `   fff   ` => color is not updated.
6. enter a valid HTML color code like `fff`=> color is updated.

### Minimal reproduction project

N/A

**Comment by guemax** [comment]
Since this seems to be an easy bug to fix (I've already dug into the source code to find the related functions), I'd like to work on it myself. I'll open a pull request once I'm happy with my approach, but the current development can be seen on [my fork of Godot](github.com/guemax/godot).

**Comment by EddieBreeg** [comment]
I would have loved to help, but that assumes we establish whether it is in fact a bug, or a design choice. I do agree it seems to be an issue that deserves to be fixed though.

**Comment by guemax** [comment]
I haven't started to make changes in my fork yet, because I wanted to wait for the reactions and opinions first. (one can see it as a problem, but it doesn't have to be.)

**Comment by guemax** [comment]
However, since other programs don't allow spaces in HTML color codes, I think it would be great to do it that way. I participated in the last GTMK Game Jam with Godot and had quite a bit of trouble copying and pasting the colors.

I managed to copy them without spaces at the end, but it was a bit tedious and not very user friendly.

**Comment by AThousandShips** [comment]
I'd argue it's a good thing to fix at least

I'd limit it to the gui, not change the behaviour in `Color`

**Comment by EddieBreeg** [comment]
The issue seems to be in the _html_submitted method of ColorPicker, as far as I can see simply creating a trimmed copy of the string should do the trick as you suggested

**Comment by AThousandShips** [comment]
The relevant code is here:
https://github.com/godotengine/godot/blob/23318e877890029f35856036b9c4e0bfa09cacc6/scene/gui/color_picker.cpp#L561-L582

**Comment by guemax** [comment]
Thanks for your advice! :smiley:  I thought about editing `Color` but it seems to have more to do with the color picker as you both pointed out.
In addition, I believe that this "wrong" input should also be fixed as early as possible, rather than deep in the source code, which should not interact with the user.

**Comment by Calinou** [comment]
> In addition, I believe that this "wrong" input should also be fixed as early as possible, rather than deep in the source code, which should not interact with the user.

Indeed, this is the [Prefer local solutions](https://docs.godotengine.org/en/stable/contributing/development/best_practices_for_engine_contributors.html#prefer-local-solutions) part of the best practices for engine contributors.

**Comment by guemax** [comment]
The PR has now been merged, and I would like to thank you all for your help, the amount of comments and advice was amazing! :smiley:
